common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -246,62 +264,188 @@ func (s *localStoreCertProvider) getClientTLSSettings(isWorker bool) *config.Cli
+	expired CertExpirationMap,
+	errors []error,
+) []error {
+
+	return fetchAndCheckCertForExpiration(
+		func() (*x509.Certificate, error) {
+			cert, err := fetchCert()
+			if err != nil {
+				return nil, err
+			}
+			if cert == nil {
+				return nil, nil
+			}
+			return x509.ParseCertificate(cert.Certificate[0])
+		},
+		time, expiring, expired, errors)
+}
 
-		caBytes, err := base64.StdEncoding.DecodeString(ca)
-		if err != nil {
-			return nil, fmt.Errorf("failed to decode ca cert: %v", err)
-		}
+func fetchAndCheckCertForExpiration(
+	fetchCert func() (*x509.Certificate, error),
+	when time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+	errors []error,
+) []error {
 
-		if !caPool.AppendCertsFromPEM(caBytes) {
-			return nil, errors.New("unknown failure constructing cert pool for ca")
-		}
+	cert, err := fetchCert()
+	if err != nil {
+		errors = append(errors, err)
+	}
+	checkCertForExpiration(cert, when, expiring, expired)
+	return errors
+}
 
-		atLeastOneCert = true
+func checkCertsForExpiration(
+	certs []*x509.Certificate,
+	time time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) {
+
+	for _, cert := range certs {
+		checkCertForExpiration(cert, time, expiring, expired)
 	}
+}
 
-	if !atLeastOneCert {
-		return nil, nil
+func checkCertForExpiration(
+	cert *x509.Certificate,
+	when time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) {
+
+	if cert != nil && cert.NotAfter.Before(when) {
+		record := CertExpirationData{
+			Thumbprint: md5.Sum(cert.Raw),
+			IsCA:       cert.IsCA,
+			DNSNames:   cert.DNSNames,
+			Expiration: cert.NotAfter,
+		}
+		if record.Expiration.Before(time.Now().UTC()) {
+			expired[record.Thumbprint] = record
+		} else {
+			expiring[record.Thumbprint] = record
+		}
 	}
+}
 
-	return caPool, nil
+func buildCAPoolFromData(caData []string) (*x509.CertPool, []*x509.Certificate, error) {
+
+	return buildCAPool(caData, base64.StdEncoding.DecodeString)
 }
 
-func buildCAPoolFromFiles(caFiles []string) (*x509.CertPool, error) {
+func buildCAPoolFromFiles(caFiles []string) (*x509.CertPool, []*x509.Certificate, error) {
+
+	return buildCAPool(caFiles, ioutil.ReadFile)
+}
+
+func buildCAPool(cas []string, getBytes func(item string) ([]byte, error)) (*x509.CertPool, []*x509.Certificate, error) {
 	atLeastOneCert := false
 	caPool := x509.NewCertPool()
+	certs := make([]*x509.Certificate, 1)
 
-	for _, ca := range caFiles {
+	for _, ca := range cas {
 		if ca == "" {
 			continue
 		}
 
-		caBytes, err := ioutil.ReadFile(ca)
+		caBytes, err := getBytes(ca)
 		if err != nil {
-			return nil, fmt.Errorf("failed to read ca cert from file '%v': %v", ca, err)
+			return nil, nil, fmt.Errorf("failed to decode ca cert: %v", err)
 		}
 
 		if !caPool.AppendCertsFromPEM(caBytes) {
-			return nil, errors.New("unknown failure constructing cert pool for ca")
+			return nil, nil, errors.New("unknown failure constructing cert pool for ca")
+		} else {
```

Comment:
given you are returning in the above if block, the "else {" is unnecessary, right?